From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: shoker537 <andreybest2002@ya.ru>
Date: Fri, 7 Jul 2023 17:53:41 +0300
Subject: [PATCH] Suppress another POI chunk expected error


diff --git a/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkFullTask.java b/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkFullTask.java
index 300700477ee34bc22b31315825c0e40f61070cd5..e2798d25f6cfa66bd745d291152fb63106cdbae4 100644
--- a/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkFullTask.java
+++ b/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkFullTask.java
@@ -43,7 +43,7 @@ public final class ChunkFullTask extends ChunkProgressionTask implements Runnabl
             // moved from the load from nbt stage into here
             final PoiChunk poiChunk = this.chunkHolder.getPoiChunk();
             if (poiChunk == null) {
-                LOGGER.error("Expected poi chunk to be loaded with chunk for task " + this.toString());
+//                LOGGER.error("Expected poi chunk to be loaded with chunk for task " + this.toString());
             } else {
                 poiChunk.load();
                 this.world.getPoiManager().checkConsistency(this.fromChunk);
