From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: shoker137 <andreybest2002@ya.ru>
Date: Sat, 18 May 2024 14:49:31 +0300
Subject: [PATCH] Synchronize when add the world to NMS worlds list


diff --git a/src/main/java/com/infernalsuite/aswm/SlimeNMSBridgeImpl.java b/src/main/java/com/infernalsuite/aswm/SlimeNMSBridgeImpl.java
index dc66d1e67d541dbd608c2b2a0e26b25cec9183ff..d31d81d422d114961abb96471b846b6de956a58a 100644
--- a/src/main/java/com/infernalsuite/aswm/SlimeNMSBridgeImpl.java
+++ b/src/main/java/com/infernalsuite/aswm/SlimeNMSBridgeImpl.java
@@ -149,9 +149,13 @@ public class SlimeNMSBridgeImpl implements SlimeNMSBridge {
             throw new IllegalArgumentException("World " + worldName + " already exists! Maybe it's an outdated SlimeWorld object?");
         }
 
-        SlimeLevelInstance server = createCustomWorld(null, slimeWorld, dimensionOverride);
-        registerWorld(server);
-        return server.getSlimeInstance();
+        SlimeLevelInstance level = createCustomWorld(null, slimeWorld, dimensionOverride);
+        if(Bukkit.isPrimaryThread()) {
+            registerWorld(level);
+        } else {
+            MinecraftServer.getServer().executeBlocking(() -> registerWorld(level));
+        }
+        return level.getSlimeInstance();
     }
 
     @Override
